components/browser/icons/src/main/java/mozilla/components/browser/icons/BrowserIcons.kt

```diff
@@ -138,6 +143,34 @@ class BrowserIcons(
                 Logger.error("Could not install browser-icons extension", throwable)
             })
     }
+
+    /**
+     * Loads an icon asynchronously using [BrowserIcons] and then displays it in the [ImageView].
+     * If the view is detached from the window before loading is completed, then loading is cancelled.
+     */
+    fun loadIntoView(view: ImageView, request: IconRequest) = scope.async(Dispatchers.Main) {
+        // If we previously started loading into the view, cancel the job.
+        val existingJob = view.getTag(R.id.mozac_browser_icons_tag_job) as? Job
+        existingJob?.cancel()
+
+        // Create a loading job
+        val deferredIcon = loadIcon(request)
+
+        view.setTag(R.id.mozac_browser_icons_tag_job, deferredIcon)
+        val onAttachStateChangeListener = CancelOnDetach(deferredIcon).also {
+            view.addOnAttachStateChangeListener(it)
+        }
+
+        try {
+            val icon = deferredIcon.await()
+            view.setImageBitmap(icon.bitmap)
+        } catch (e: CancellationException) {
+            view.setImageDrawable(null)
```

Comment:
I wonder if we really want to set the drawable to `null` here? 🤔

In the scenario of a RecyclerView that would be the job of the adapter. In other scenarios inside the app the app may actually set a placeholder image that we would temporarily clear here until a new icon is loaded.

Reply:
I'll add parameters for a placeholder and failure drawable and default to null.